pkg/api/volume/formatter.go

```diff
@@ -22,21 +30,34 @@ func Formatter(request *types.APIRequest, resource *types.RawResource) {
 		return
 	}
 
-	if canExport(pvc) {
-		resource.AddAction(request, actionExport)
+	if IsResizing(pvc) {
+		resource.AddAction(request, actionCancelExpand)
+		return
 	}
 
-	if canCancelExpand(pvc) {
-		resource.AddAction(request, actionCancelExpand)
+	if pvc.Status.Phase != corev1.ClaimBound {
+		return
 	}
-}
 
-func canExport(pvc *corev1.PersistentVolumeClaim) bool {
-	return !IsResizing(pvc)
-}
+	csiDriverConfig := make(map[string]settings.CSIDriverInfo)
+	if err := json.Unmarshal([]byte(settings.CSIDriverConfig.Get()), &csiDriverConfig); err != nil {
+		return
+	}
+	provisioner := pvc.Annotations[util.AnnotationVolumeKubernetesProvisioner]
```

Comment:
How about getting the provisioner from `pvc.spec.storageClassName` and then `storageClass.provisioner`? We can update `Formatter` function like [vmformatter](https://github.com/harvester/harvester/blob/master/pkg/api/vm/formatter.go), so we can have a related cache to get StorageClass.

Reply:
why not getting the provisioner from annotation directly?

Reply:
I don't think we can make sure all PVC has this annotation.

Reply:
external-provisioner uses different annotations in different versions.

https://github.com/kubernetes-csi/external-provisioner/blob/afca22d79f718300f851de0d8912330cf8b8b04a/pkg/controller/controller.go#L136

https://github.com/kubernetes-csi/external-provisioner/blob/fdfd8a39fce5391f7e8737cf3c373b118ef40957/pkg/controller/controller.go#L131-L132

Reply:
added a common function `GetPVCProvisioner` https://github.com/harvester/harvester/pull/2648/files#diff-2d04e14ba508495210e795a4e38c21cb65879de10f4fe33ce571f684011c00b9

Reply:
added a common function `GetPVCProvisionerFromAnnotation`  https://github.com/harvester/harvester/pull/2648/files#diff-2d04e14ba508495210e795a4e38c21cb65879de10f4fe33ce571f684011c00b9


```go
import (
	corev1 "k8s.io/api/core/v1"
)

const (
	AnnStorageProvisioner     = "volume.kubernetes.io/storage-provisioner"
	AnnBetaStorageProvisioner = "volume.beta.kubernetes.io/storage-provisioner"
)

// GetPVCProvisionerFromAnnotation do not use this function when the PVC is just created
func GetPVCProvisionerFromAnnotation(pvc *corev1.PersistentVolumeClaim) string {
	provisioner, ok := pvc.Annotations[AnnBetaStorageProvisioner]
	if !ok {
		provisioner = pvc.Annotations[AnnStorageProvisioner]
	}
	return provisioner
}

```

Reply:
we can make sure a provisioned PVC has these annotations always （these annotations seted by pv controller before pv provisioned）， refer to https://github.com/kubernetes/kubernetes/blob/90f9a52db6de21cafebb91a7f47b7386f37289e0/pkg/controller/volume/persistentvolume/pv_controller.go#L1594

Reply:
added a common function `GetPVCProvisionerFromAnnotation`  https://github.com/harvester/harvester/pull/2648/files#diff-2d04e14ba508495210e795a4e38c21cb65879de10f4fe33ce571f684011c00b9


```go
import (
	corev1 "k8s.io/api/core/v1"
)

const (
	AnnStorageProvisioner     = "volume.kubernetes.io/storage-provisioner"
	AnnBetaStorageProvisioner = "volume.beta.kubernetes.io/storage-provisioner"
)

// GetProvisionedPVCProvisioner do not use this function when the PVC is just created
func GetProvisionedPVCProvisioner(pvc *corev1.PersistentVolumeClaim) string {
	provisioner, ok := pvc.Annotations[AnnBetaStorageProvisioner]
	if !ok {
		provisioner = pvc.Annotations[AnnStorageProvisioner]
	}
	return provisioner
}

```